doc/contributing.rst

```diff
@@ -1,221 +1,577 @@
-* Avoid unnecessary cosmetic changes in PRs.
-* Minimize test timing while maximizing coverage. Use ``pytest --durations=20`` on modified tests.
-* Update the ``doc/whats_new.rst`` file last, just before merge to avoid merge conflicts.
+.. code-block:: console
 
-Naming
-^^^^^^
-* Classes should be named using CamelCase.
-* Functions and instances/variables should be snake_case (n_samples rather than nsamples).
-* Avoid single-character variable names.
+    $ pip uninstall -y mne  # or: conda remove --force mne
+    $ pip install -e .      # or: python setup.py develop
 
-Importing
-^^^^^^^^^
-* Import modules in this order:
-  1. builtin
-  2. standard scientific (``numpy as np``, ``scipy`` submodules)
-  3. others
-  4. mne imports (relative within the MNE module, absolute in the examples)
-* Imports for ``matplotlib`` and optional modules (``sklearn``, and ``pandas``, etc.) within the MNE module should be nested (i.e., within a function or method, not at the top of a file).
+Next, we'll do the same thing for Mayavi (one of our 3D plotting backends):
 
-Vizualization
-^^^^^^^^^^^^^
-* Add public functions to the :mod:`mne.viz` package and use these in the corresponding methods.
-* All visualization functions must accept a ``show`` parameter and return a ``fig`` handle.
-* Use ``RdBu_r`` colormap for signed data with a meaningful middle (zero-point) and ``Reds`` otherwise in visualization functions and examples.
+.. code-block:: console
 
-Return types
-^^^^^^^^^^^^
-* Methods should modify inplace and return ``self``, functions should return copies (where applicable).
-
-Style
-^^^^^
-* Use single quotes whenever possible.
-* Prefer generator or list comprehensions over ``filter``, ``map`` and other functional idioms.
-* Use explicit functional constructors for builtin containers to improve readability (e.g., ``list()``, ``dict``).
-* Avoid nested functions or class methods if possible -- use private functions instead.
-* Avoid ``**kwargs`` and ``*args`` in function signatures.
-* Add brief docstrings to simple private functions and complete docstrings for complex ones.
-
-Checking and building documentation
------------------------------------
-
-All changes to the codebase must be properly documented.
-To ensure that documentation is rendered correctly, the best bet is to
-follow the existing examples for class and function docstrings,
-and examples and tutorials.
-
-Our documentation (including docstring in code) uses ReStructuredText format,
-see `Sphinx documentation`_ to learn more about editing them. Our code
-follows the `NumPy docstring standard`_.
-
-Documentation is automatically built remotely during pull requests. If
-you want to also test documentation locally, you will need to install (using
-``conda`` or ``pip`` as preferred):
+    $ pip uninstall -y mayavi
+    $ cd $INSTALL_LOCATION
+    $ git clone git://github.com/enthought/mayavi.git
+    $ cd mayavi
+    $ pip install -e .  # or: python setup.py develop
```

Comment:
The stable version of mayavi seems to work just fine for me. Why would you need to do this?

Reply:
May no longer be needed... I think it was something about a regex in mayavi making master of Sphinx gallery unhappy (unrecognized escape sequence)?

Reply:
It was fixed 5 days ago:

https://github.com/enthought/mayavi/pull/747/files

Let's just give the `pip` install line that installs `master` for such dependencies

Reply:
oops.  I missed this comment before pushing the revisions.  @larsoner are you okay with what I did (installing dependencies via env file or pip, with a paragraph about updating dependencies to master when needed), or should I switch it to install them from master by default?  (and if so, specifically which dependencies should get installed from master?)